### PR TITLE
Smar 72 notification 5 min

### PR DESCRIPTION
**Pull request by yulyaolshanska**
## Description changes:

- Notification(in 5 min). Frontend development.
- Connect front and back (Notification next meeting in 5 min) ;

### General
- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [ ] Updated the documentation
- [x] Performed a self-review of my code
- [x] Don't have "any" on my code
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside ternary operator
- [x] Don't have commented code
- [x] All links in on ENV files
- [x] Used camelCase for variables and functions
- [x] Functions is public only if it's used outside the class
- [x] No hardcoded values
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Frontend
- [x] Components and business logic are separated
- [x] Colors, Font Size, and Font Name is on the theme or in the constants
- [x] No text in the components, use i18n approach
- [x] No inline styles
- [x] Imports are absolute
- [x] Attach a screenshot if PR has visual changes.

### Screenshots
Timer 5 min before meeting

image

When the meeting started

image

### CODE REVIEW

File: src/components/MeetNotification/index.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  const diffTimeEnd =
+    new Date(endTime).getTime() - new Date(formattedCurrentTime).getTime();
+  const isTimerRun = diffTimeStart < fiveMinutes && diffTimeStart > sec;
+  const isMeetInProgress = diffTimeEnd > 0 && diffTimeEnd <= fiveMinutes;
+  const isNotifOpen = isTimerRun || isMeetInProgress;
+
+  const deadline = new Date(startTime);
+  const diff = deadline.getTime() - new Date().getTime() || 0;
+  const minutes = diff > 0 ? Math.floor(timer / 60) % 60 : 0;
+  const seconds = diff > 0 ? Math.floor(timer % 60) : 0;
```
Comment (lyudad) [inline src/components/MeetNotification/index.tsx]: наша песня нованачинаем ее снова, что это? мы не jquery пишем. эта вся бадя будет при каждом перерендере пересчитыватся + зачем ты деблируешь финкции почему не написать красивый врапер. не, это выкинуть и принести красивый код

**Review by SwiftXLight — APPROVED** [review]

File: src/components/MeetNotification/index.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+type TimerResult = {
+  minutes: number;
+  seconds: number;
+};
+
+const MeetNotification = () => {
+  const { t } = useTranslation();
+  const dispatch = useAppDispatch();
+  const [timer, setTimer] = useState(0);
```
Comment (lyudad) [inline src/components/MeetNotification/index.tsx]: type

File: src/components/MeetNotification/index.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+  const { patient, startTime, endTime, remoteDoctor, localDoctor } =
+    useAppSelector((state) => state.socketAppointmentReducer.nextAppointment);
+  const doctor = useAppSelector((state) => state.doctorReducer);
+
+  const formattedCurrentTime = useMemo(
```
Comment (lyudad) [inline src/components/MeetNotification/index.tsx]: it look like could be a general function for full project not only for current component, its mean it should be moved to, for example, `utility` folder in the root folder

File: src/components/MeetNotification/index.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+  const isMeetInProgress = useMemo(
+    () => diffTimeEnd > 0 && diffTimeEnd <= fiveMinutes,
+    [diffTimeEnd]
+  );
+
+  const isNotifOpen = useMemo(
+    () => isTimerRun || isMeetInProgress,
+    [isTimerRun, isMeetInProgress]
+  );
```
Comment (lyudad) [inline src/components/MeetNotification/index.tsx]: I accept useMemo, but you need to learn react, useMemo is bad approach in this case

File: src/components/MeetNotification/index.tsx [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+  useEffect(() => {
+    const diffEnd = new Date(endTime).getTime() - new Date().getTime();
```
Comment (lyudad) [inline src/components/MeetNotification/index.tsx]: you have same function diffTimeEnd on line 56, why is duplication needed? why you not create one generic function which will accept time for calculation

File: src/components/MeetNotification/index.tsx [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  const getNotificationText = useMemo(() => {
+    return isTimerRun
+      ? `${t('Notification.youHaveVideoCallIn')}  ${`${minutes
+          .toString()
+          .padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`} minutes
+            with:`
+      : `${t('Notification.youCurrentlyHaveMeeting')}`;
+  }, [minutes, seconds]);
+
+  return (
```
Comment (lyudad) [inline src/components/MeetNotification/index.tsx:100]: you have 162 lines of logic calculation in the component? don't you think its a good sign to create custom hook

**Review by OlenaPosternak — APPROVED** [review]

File: src/components/Drawer/index.tsx [hunk 7]
```diff
@@ -83,14 +84,20 @@ const Drawer = () => {
 
   const cancelLogout = () => {
     setShowModal(false);
-  }
+  };
+
+  const getDoctorSpecialization = useMemo(() => {
+    return currentDoctor.specialization === 1
```
Comment (lyudad) [inline src/components/Drawer/index.tsx]: magic number

File: src/components/Zoom/socket-io.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+const socketIOUrl = `${import.meta.env.VITE_REACT_APP_BASE_URL_SERVER}${
+  import.meta.env.VITE_APPOINTMENT_NAMESPACE
+}`;
+
+const dispatch = store.dispatch;
+
+export const createSocketWithHandlers = (): Socket => {
+  const token = cookie.get('accessToken');
+  console.log(`Creating socket with token: ${token}`);
```
Comment (lyudad) [inline src/components/Zoom/socket-io.ts]: what is it?

File: src/components/Zoom/socket-io.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    },
+    transports: ['websocket', 'polling'],
+  });
+
+  socket.on('connect', () => {
+    console.log(`Connected with socket ID: ${socket.id}.`);
+  });
+
+  socket.on('appointment_update', (nextAppointment) => {
+    console.log('event: appointment_update', nextAppointment);
```
Comment (lyudad) [inline src/components/Zoom/socket-io.ts]: ?

File: src/utils/hooks/useTimer.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  seconds: number;
+};
+
+const useTimer = (startTime: string): TimerData => {
+  const [timer, setTimer] = useState<number>(0);
+  const minutes = Math.floor(timer / SEC_PER_MIN) % SEC_PER_MIN;
+  const seconds = Math.floor(timer % SEC_PER_MIN);
+
+  useEffect(() => {
+    const diffTimeStart =
```
Comment (lyudad) [inline src/utils/hooks/useTimer.ts]: I still dont understand why you didnt make it general function

**Review by OlenaPosternak — APPROVED** [review]

File: src/components/MeetNotification/index.tsx [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+import { local } from '@constants/other';
+import { female } from '@constants/patient';
+import { fiveMinutes } from '@constants/notification';
+import useTimer from 'utils/hooks/useTimer';
+import cookie from 'utils/functions/cookies';
+import useNotificationText from 'utils/hooks/useNotificationText';
+import { getCurrentFormattedTime, getDiffTime } from 'utils/functions/timeUtils';
+import { notificationCurrentTime } from '@constants/format';
+
+const token = cookie.get('accessToken');
```
Comment (lyudad) [inline src/components/MeetNotification/index.tsx:26]: почему токен в куки? для этого есть redux persist

File: src/components/MeetNotification/styles.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  color: ${ZAMBEZI};
+`;
+
+export const DetailsBtn = styled(NavLink)`
+  padding: 12px 24px;
+  height: 44px;
+  background: ${NOTIFICATION_BTN_BACKGROUND};
+  border-radius: 4px;
+  font-weight: 700;
+  font-size: 14px;
```
Comment (lyudad) [inline src/components/MeetNotification/styles.ts:35]: ?

File: src/components/MeetNotification/styles.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+`;
+
+export const CallInfo = styled.div`
+  display: flex;
+`;
+
+export const El = styled.span`
+  margin: 0 5px;
+  font-weight: 400;
+  font-size: 16px;
```
Comment (lyudad) [inline src/components/MeetNotification/styles.ts:65]: ?

File: src/utils/hooks/useTimer.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      setTimer(initialTimer);
+    }
+
+    return () => {
+      clearInterval(intervalId);
+    };
+  }, [startTime]);
+
+  return { timer, minutes, seconds };
+};
```
Comment (lyudad) [inline src/utils/hooks/useTimer.ts:33]: это совсем не поняла. почему юз ефект в utils

**Pull request merged**